File: examples/canonical-analytics/python/src/analytics_workbench/analytics/incidents.py
from dataclasses import dataclass
from ..data.models import Incident, Service
from ..filters import FilterState

@dataclass
class IncidentSeverity:
    severities: list[str]
    counts: list[int]
    mttr_by_severity: dict[str, float]

def aggregate_incident_severity(incidents: list[Incident], filters: FilterState, services: list[Service] = None) -> IncidentSeverity:
    import pandas as pd
    
    if services is None:
        from ..data.loader import get_dataset
        services = get_dataset().services
    
    valid_services = {s.name for s in services}
    
    filtered = []
    for i in incidents:
        d = i.timestamp[:10]
        if filters.date_from and d < filters.date_from.isoformat(): continue
        if filters.date_to and d > filters.date_to.isoformat(): continue
        if filters.environment and i.environment not in filters.environment: continue
        if filters.severity and i.severity not in filters.severity: continue
        
        if i.service not in valid_services: continue
        if filters.services and i.service not in filters.services: continue
        
        filtered.append({
            "severity": i.severity,
            "mttr": i.mttr_mins
        })
        
    if not filtered:
        return IncidentSeverity(severities=[], counts=[], mttr_by_severity={})
        
    df = pd.DataFrame(filtered)
    
    # Order severities
    order = ["critical", "high", "medium", "low"]
    
    agg = df.groupby("severity").agg(
        count=("severity", "count"),
        mttr=("mttr", "mean")
    )
    
    # reindex
    present_order = [s for s in order if s in agg.index]
    agg = agg.loc[present_order]
    
    mttr_dict = {row.Index: float(row.mttr) if pd.notna(row.mttr) else 0.0 for row in agg.itertuples()}
    
    return IncidentSeverity(
        severities=agg.index.tolist(),
        counts=agg["count"].tolist(),
        mttr_by_severity=mttr_dict
    )

File: examples/canonical-analytics/elixir/analytics_workbench/lib/analytics_workbench/analytics/incidents.ex
defmodule AnalyticsWorkbench.Analytics.Incidents do
  alias AnalyticsWorkbench.Filters

  def aggregate_incident_severity(incidents, %Filters{} = filters) do
    filtered = apply_filters(incidents, filters)

    severities = ["sev1", "sev2", "sev3"]
    
    counts = Enum.map(severities, fn s ->
      Enum.count(filtered, & to_string(&1.severity) == s)
    end)

    mttr_by_severity = Enum.map(severities, fn s ->
      s_incidents = Enum.filter(filtered, & to_string(&1.severity) == s)
      durations = Enum.map(s_incidents, & &1.mttr_mins)
      avg = if length(durations) > 0, do: Enum.sum(durations) / length(durations), else: 0.0
      {s, Float.round(avg, 2)}
    end) |> Map.new()

    %{severities: severities, counts: counts, mttr_by_severity: mttr_by_severity}
  end

  defp apply_filters(incidents, filters) do
    incidents
    |> filter_by_date(filters)
    |> filter_by_multi(:service, filters.services)
    |> filter_by_multi(:environment, filters.environment)
    |> filter_by_multi(:severity, filters.severity)
  end

  defp filter_by_date(events, %{date_from: from, date_to: to}) do
    events
    |> Enum.filter(fn e ->
      (is_nil(from) or from == "" or String.slice(e.timestamp, 0, 10) >= from) and
      (is_nil(to) or to == "" or String.slice(e.timestamp, 0, 10) <= to)
    end)
  end

  defp filter_by_multi(events, _key, []), do: events
  defp filter_by_multi(events, key, values) do
    Enum.filter(events, &(to_string(Map.get(&1, key)) in values))
  end
end
